components/runners/ambient-runner/ambient_runner/bridges/claude/bridge.py

```
@@ -121,6 +123,22 @@ async def run(self, input_data: RunAgentInput) -> AsyncIterator[BaseEvent]:
 
+            # Capture halt state for this thread to avoid race conditions
+            # with concurrent runs modifying the shared adapter's halted flag
+            self._halted_by_thread[thread_id] = self._adapter.halted
+
+            # If the adapter halted (frontend tool or built-in HITL tool like
+            # AskUserQuestion), interrupt the worker to prevent the SDK from
+            # auto-approving the tool call with a placeholder result.
+            if self._halted_by_thread.get(thread_id, False):
+                logger.info(
+                    f"Adapter halted for thread={thread_id}, "
+                    "interrupting worker to await user input"
```

Ruff (0.15.5)

[warning] Logging statement uses f-string

(G004)